src/zenml/mlstacks/utils.py

```diff
@@ -0,0 +1,525 @@
+    }
+
+    def _add_config(
+        component_metadata: ComponentMetadata, keys: List[str]
+    ) -> None:
+        """Adds key-value pair to the component_metadata config if it exists in extra_config.
+
+        Args:
+            component_metadata: The metadata of the component.
+            keys: The keys of the configurations.
+        """
+        for key in keys:
+            value = extra_config.get(key)
+            if value is not None:
+                component_metadata.config[key] = value
+
+    # Define a list of component attributes that need checking
+    component_attributes = ["component_type", "component_flavor", "provider"]
+
+    for component in components:
+        component_metadata = ComponentMetadata()
+        component_metadata.config = {}
+
+        # Create a dictionary that maps attribute names to their values
+        component_values = {
+            attr: getattr(component, attr, None)
+            for attr in component_attributes
+        }
+
+        for config_keys, extra_keys in config_map.items():
+            # If all attributes in config_keys match their values in the component
+            if all(
+                component_values.get(key) == value
+                for key, value in zip(component_attributes, config_keys)
+            ):
+                _add_config(component_metadata, extra_keys)
+
+        # always add project_id to gcp components
+        if component.provider == "gcp":
+            project_id = extra_config.get("project_id")
+            if project_id:
+                component_metadata.config["project_id"] = extra_config.get(
+                    "project_id"
+                )
+            else:
+                raise KeyError(
+                    "No `project_id` is included. Please try again with "
+                    "`--extra-config project_id=<project_id>`"
+                )
+
+        component.metadata = component_metadata
+
+    return components
+
+
+@verify_installation
+def _construct_components(
+    params: Dict[str, Any], zenml_component_deploy: bool = False
+) -> List["Component"]:
+    """Constructs mlstacks `Component` objects from raw Click CLI params.
+
+    Args:
+        params: Raw Click CLI params.
+        zenml_component_deploy: A boolean indicating whether the stack
+            contains a single component and is being deployed through `zenml
+            component deploy`
+
+    Returns:
+        A list of mlstacks `Component` objects.
+    """
+    from mlstacks.models import Component
+
+    provider = params["provider"]
+    extra_config = (
+        dict(config.split("=") for config in params["extra_config"])
+        if params["extra_config"]
+        else {}
+    )
+    if zenml_component_deploy:
+        components = [
+            Component(
+                name=params[key],
+                component_type=key,
+                component_flavor=_get_component_flavor(key, value, provider),
```

Comment:
No, because the flavors are sometimes bools and sometimes flavors etc. If you look at the logic of the `_get_component_flavor` method you'll see why we can't just get the flavor straight.